sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala

```diff
@@ -421,48 +421,364 @@ case class Cast(child: Expression, dataType: DataType) extends UnaryExpression w
 
   protected override def nullSafeEval(input: Any): Any = cast(input)
 
-  override def genCode(ctx: CodeGenContext, ev: GeneratedExpressionCode): String = {
-    // TODO: Add support for more data types.
-    (child.dataType, dataType) match {
+  override def genCode(ctx: CodeGenContext, ev: GeneratedExpressionCode): String =
+    castGen(child.dataType, dataType, ctx, ev)
 
-      case (BinaryType, StringType) =>
-        defineCodeGen (ctx, ev, c =>
-          s"${ctx.stringType}.fromBytes($c)")
+  private[this] def castGen(from: DataType, to: DataType, ctx: CodeGenContext,
+      ev: GeneratedExpressionCode): String = {
 
-      case (DateType, StringType) =>
+    def castToStringCode(from: DataType): String = from match {
```

Comment:
it'd be better to put these as top level functions instead of nested functions since castGen is already super long